packages/core/src/services/commits/runDocumentAtCommit.ts

```diff
@@ -5,57 +5,92 @@ import {
   ChainEventTypes,
   StreamEventTypes,
 } from '@latitude-data/core'
-import type { Commit, ProviderApiKey } from '@latitude-data/core/browser'
-import { findWorkspaceFromCommit } from '$core/data-access'
-import { NotFoundError, Result } from '$core/lib'
+import type {
+  Commit,
+  DocumentVersion,
+  ProviderApiKey,
+  Workspace,
+} from '@latitude-data/core/browser'
+import { NotFoundError, Result, UnprocessableEntityError } from '$core/lib'
 import { streamToGenerator } from '$core/lib/streamToGenerator'
 import { ProviderApiKeysRepository } from '$core/repositories'
+import { v4 as uuid } from 'uuid'
+import { ZodError } from 'zod'
 
 import { ai, AILog, validateConfig } from '../ai'
 import { getResolvedContent } from '../documents/getResolvedContent'
 
+type CachedApiKeys = Map<string, ProviderApiKey>
+async function cacheApiKeysByName({ workspaceId }: { workspaceId: number }) {
+  const scope = new ProviderApiKeysRepository(workspaceId)
+  const result = await scope.findAll().then((r) => r.unwrap())
+
+  return result.reduce((acc, apiKey) => {
+    acc.set(apiKey.name, apiKey)
+    return acc
+  }, new Map<string, ProviderApiKey>())
+}
+
 export async function runDocumentAtCommit({
-  documentUuid,
+  workspaceId,
+  document,
   commit,
   parameters,
-  logHandler,
+  providerLogHandler,
+  generateUUID = uuid,
 }: {
-  documentUuid: string
+  workspaceId: Workspace['id']
+  document: DocumentVersion
   commit: Commit
   parameters: Record<string, unknown>
-  logHandler: (log: AILog) => void
+  providerLogHandler: (log: AILog) => void
+  generateUUID?: typeof uuid
 }) {
-  const workspace = await findWorkspaceFromCommit(commit)
```

Comment:
Now we don't pass workspace because is passed in the API gateway. I know we should pass object to service and the external layer validates ownership (which is already doing). But this service is already doing their own check inside `getResolvedContent` so this was a redundant Db call